replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -14,139 +24,83 @@ private[journal] object Batch {
 
   def of(records: NonEmptyList[ActionRecord[Action]]): List[Batch] = {
 
-    // returns `true` when we can optimize Cassandra usage by NOT inserting "previous" Append actions in DB
-    // because "current" Delete action will discard them all
-    def cut(appends: Appends, delete: Action.Delete): Boolean = {
-      val append = appends.records.head.action
-      append.range.to <= delete.to.value
-    }
-
-    def dropDeleted(appends: Appends, delete: Action.Delete): Option[Appends] = {
-      // we cannot drop individual events in record as we do not parse the payload here (`records.head.action.payload: ByteVector`)
-      val validRecords = appends.records.collect { case record if record.action.range.to >= delete.to.value => record }
-      NonEmptyList.fromList(validRecords).map { Appends(appends.offset, _) }
-    }
-
-    records
-      .foldLeft(List.empty[Batch]) { (bs, record) =>
-        val offset = record.partitionOffset.offset
-
-        def appendsOf(records: NonEmptyList[ActionRecord[Action.Append]]): Appends = {
-          Appends(offset, records)
+    val state = records.reverse.foldLeft(State()) {
+      // ignore all actions before `Purge`
+      case (state, _) if state.purge.nonEmpty =>
+        state
+
+      case (state, ActionRecord(purge: Action.Purge, po)) =>
+        state.copy(
+          purge = Purge(po.offset, purge.origin, purge.version).some,
+        )
+
+      case (state, ActionRecord(delete: Action.Delete, po)) =>
+        val delete_ = state.delete match {
+          case Some(younger) =>
+            // take `origin` and `version` from "older" entity, if it has them
+            val origin  = delete.origin.orElse(younger.origin)
+            val version = delete.version.orElse(younger.version)
+            // make `Delete` action with largest `seqNr` and largest `offset`
+            if (younger.to < delete.to) Delete(po.offset, delete.to, origin, version)
+            else younger.copy(origin = origin, version = version)
+
+          case None =>
+            Delete(po.offset, delete.to, delete.origin, delete.version)
         }
-
-        def deleteOf(to: DeleteTo, origin: Option[Origin], version: Option[Version]): Delete = {
-          Delete(offset, to, origin, version)
+        state.copy(
+          delete = delete_.some,
+        )
+
+      case (state, ActionRecord(append: Action.Append, po)) if state.delete.forall(_.to.value <= append.range.to) =>
```

Comment:
I wonder if these `Delete` actions could come in a wrong order? Or that cannot happen?

I.e. like this:
```
Delete(3)
Append(1)
Append(2)
Append(3)
```
In this case, after `records.reverse` that `Delete(3)` will be handled last, and the appends will not be cleaned up.

Reply:
passing test-case:
```scala
      (
        Nel.of(
          delete(offset = 1, to = 3),
          append(offset = 2, seqNr = 1),
          append(offset = 3, seqNr = 2),
          append(offset = 4, seqNr = 3),
        ),
        List(
          appends(offset = 4, append(offset = 2, seqNr = 1), append(offset = 3, seqNr = 2), append(offset = 4, seqNr = 3)),
          deletes(offset = 4, to = 3),
        ),
      ),
```
it would add and delete those events - sadly, they are not removed by `Batch`.

in original code, to pass test-case, the check should be:
```scala
        List(
          deletes(offset = 1, to = 3),
          appends(offset = 4, append(offset = 2, seqNr = 1), append(offset = 3, seqNr = 2), append(offset = 4, seqNr = 3)),
        ),
```

indeed, this creates inconsistency with original handling, as `Delete(3)` attempts to delete "in future" and in original version the events are "leaked" in `journal` table, while in this PR they will get added and removed.

if it would get applied in 2 batches - `Delete(3)` and then all `Append(*)`, then:
* in `metajournal` we would create head with `delete_to=3, seqNr = 3` (this is valid way to create new journal)
* all `Append(*)` would get ignored - those events would never become available for journal

when we apply them all in single batch, then:
* in `metajournal` we expect to create head with `delete_to=3, seqNr = 3`
* there should not be events in `journal` table - here the original code "leaks" them, while new one will cleanup

overall - such "delete in future" is not expected, though with bug in code it is possible. I am not sure, if there is an easy way how to overcome it - there is no way for `kafka-journal` to provide feedback to client about inconsistent or unexpected behaviour. 

test-case in original code for `ReplicatedCassandraTest` demonstrates that events are "leaked" (on recovery they will be ignored):
```scala
    test(s"batched `[delete(2), append(1), append(2)]` must update tables correctly, $suffix") {
      val id          = "id"
      val key         = Key(id = id, topic = topic0)
      val metaSegment = segmentOfId.metaJournal(key)

      val stateT = for {
        _ <- journal.delete(
          key       = key,
          partition = Partition.min,
          offset    = Offset.unsafe(1),
          timestamp = timestamp0,
          deleteTo  = SeqNr.unsafe(2).toDeleteTo,
          origin    = none,
        )
        _ <- journal.append(
          key         = key,
          partition   = Partition.min,
          offset      = Offset.unsafe(2),
          timestamp   = timestamp1,
          expireAfter = none,
          events = Nel.of(
            eventRecordOf(seqNr = SeqNr.unsafe(1), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(2))).event,
          ),
        )
        _ <- journal.append(
          key         = key,
          partition   = Partition.min,
          offset      = Offset.unsafe(3),
          timestamp   = timestamp1,
          expireAfter = none,
          events = Nel.of(
            eventRecordOf(seqNr = SeqNr.unsafe(2), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(3))).event,
          ),
        )
      } yield {}

      val expected = State(
        actions = List(
          Action.UpdateSeqNr(key, metaSegment, PartitionOffset(Partition.min, Offset.unsafe(3)), timestamp1, SeqNr.unsafe(2)),
          Action.InsertRecords(key, SegmentNr.journal(SeqNr.unsafe(1), segmentSize), 1),
          Action.UpdateSeqNr(key, metaSegment, PartitionOffset(Partition.min, Offset.unsafe(2)), timestamp1, SeqNr.unsafe(1)),
          Action.InsertRecords(key, SegmentNr.journal(SeqNr.unsafe(2), segmentSize), 1),
          Action.InsertMetaJournal(
            key,
            metaSegment,
            created = timestamp0,
            updated = timestamp0,
            journalHead = JournalHead(
              partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(1)),
              segmentSize     = segmentSize,
              seqNr           = SeqNr.unsafe(2),
              deleteTo        = SeqNr.unsafe(2).toDeleteTo.some,
            ),
            origin = none,
          ),
        ),
        metaJournal = Map(
          (
            (topic0, metaSegment),
            Map(
              (
                id,
                MetaJournalEntry(
                  journalHead = JournalHead(
                    partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(3)),
                    segmentSize     = segmentSize,
                    seqNr           = SeqNr.unsafe(2),
                    deleteTo        = SeqNr.unsafe(2).toDeleteTo.some,
                    expiry          = none,
                    recordId        = none,
                  ),
                  created = timestamp0,
                  updated = timestamp1,
                  origin  = none,
                ),
              ),
            ),
          ),
        ),
        journal = Map( // <=== from perspective of `metajournal` table, this is expected to be `empty` - leak
          (
            (key, SegmentNr.journal(SeqNr.unsafe(1), segmentSize)),
            Map(
              (
                (SeqNr.unsafe(1), timestamp0),
                eventRecordOf(seqNr = SeqNr.unsafe(1), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(2)))
                  .copy(metaRecordId = none),
              ),
              (
                (SeqNr.unsafe(2), timestamp0),
                eventRecordOf(seqNr = SeqNr.unsafe(2), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(3)))
                  .copy(metaRecordId = none),
              ),
            ),
          ),
        ),
      )
      val result = stateT.run(State.empty)
      result shouldEqual (expected, ()).pure[Try]
    }
```

Reply:
honestly, I am not sure what is better:
* leaking events in original code
* adding and removing them in code as it is now in this PR

the consistency is better now, though load on Cassandra is not optimal

Reply:
hm.. `ReplicatedCassandra` doesn't check if `seqNr` of new event is before `deleteTo` for this journal... :garold:

Reply:
following test-case for actions `Delete(12), Append(11), Append(12)` on head with `deleteTo: 10, seqNr:10` "leaks" journal events: `11`, `12` and fails to update `deleteTo` in head - it stays at initial value `10` - code in `ReplicatedCassandra` doesn't allow "delete in future" and clamps attempt down to `seqNr`.

<details>
<summary></summary>

```scala
    test(s"batched `[delete(12), append(11), append(12)]` on active heat must update tables correctly, $suffix") {
      val id          = "id"
      val key         = Key(id = id, topic = topic0)
      val metaSegment = segmentOfId.metaJournal(key)

      val initial = State(
        metaJournal = Map(
          (
            (topic0, metaSegment),
            Map(
              (
                id,
                MetaJournalEntry(
                  journalHead = JournalHead(
                    partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(11)),
                    segmentSize     = segmentSize,
                    seqNr           = SeqNr.unsafe(10L),
                    deleteTo        = SeqNr.unsafe(10L).toDeleteTo.some,
                    expiry          = none,
                    recordId        = none,
                  ),
                  created = timestamp0,
                  updated = timestamp0,
                  origin  = none,
                ),
              ),
            ),
          ),
        ),
      )

      val stateT = for {
        _ <- journal.delete(
          key       = key,
          partition = Partition.min,
          offset    = Offset.unsafe(12),
          timestamp = timestamp1,
          deleteTo  = SeqNr.unsafe(12).toDeleteTo,
          origin    = none,
        )
        _ <- journal.append(
          key         = key,
          partition   = Partition.min,
          offset      = Offset.unsafe(13),
          timestamp   = timestamp1,
          expireAfter = none,
          events = Nel.of(
            eventRecordOf(seqNr = SeqNr.unsafe(11), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(13))).event,
          ),
        )
        _ <- journal.append(
          key         = key,
          partition   = Partition.min,
          offset      = Offset.unsafe(14),
          timestamp   = timestamp1,
          expireAfter = none,
          events = Nel.of(
            eventRecordOf(seqNr = SeqNr.unsafe(12), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(14))).event,
          ),
        )
      } yield {}

      val expected = State(
        actions = List(
          Action.UpdateSeqNr(key, metaSegment, PartitionOffset(Partition.min, Offset.unsafe(14)), timestamp1, SeqNr.unsafe(12)),
          Action.InsertRecords(key, SegmentNr.journal(SeqNr.unsafe(12), segmentSize), 1),
          Action.UpdateSeqNr(key, metaSegment, PartitionOffset(Partition.min, Offset.unsafe(13)), timestamp1, SeqNr.unsafe(11)),
          Action.InsertRecords(key, SegmentNr.journal(SeqNr.unsafe(11), segmentSize), 1),
          Action.UpdatePartitionOffset(key, metaSegment, PartitionOffset(Partition.min, Offset.unsafe(12)), timestamp1),
        ),
        metaJournal = Map(
          (
            (topic0, metaSegment),
            Map(
              (
                id,
                MetaJournalEntry(
                  journalHead = JournalHead(
                    partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(14)),
                    segmentSize     = segmentSize,
                    seqNr           = SeqNr.unsafe(12),
                    deleteTo        = SeqNr.unsafe(10).toDeleteTo.some,
                    expiry          = none,
                    recordId        = none,
                  ),
                  created = timestamp0,
                  updated = timestamp1,
                  origin  = none,
                ),
              ),
            ),
          ),
        ),
        journal = Map(
          (
            (key, SegmentNr.journal(SeqNr.unsafe(12), segmentSize)),
            Map(
              (
                (SeqNr.unsafe(11), timestamp0),
                eventRecordOf(seqNr = SeqNr.unsafe(11), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(13)))
                  .copy(metaRecordId = none),
              ),
              (
                (SeqNr.unsafe(12), timestamp0),
                eventRecordOf(seqNr = SeqNr.unsafe(12), partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(14)))
                  .copy(metaRecordId = none),
              ),
            ),
          ),
        ),
      )
      val result = stateT.run(initial)
      result shouldEqual (expected, ()).pure[Try]
    }
```
</details>

Reply:
the existing code with batch `Delete(12), Append(11),  Append(12)` would:
* for existing journal:
** leave head with `deleteTo: 10, seqNr: 12`
** "leak" 2 events
* for non-existing journal:
** create journal with `deleteTo: 12, seqNr: 12`
** "leak" 2 events without updating `metajournal` table

new code would optimize `Delete(12), Append(11),  Append(12)` to `Append(11),  Append(12), Delete(12)` and:
* for both existing journal and non-existing one:
** update head with `deleteTo: 12, seqNr: 12`
** add and remove event `12`